contracts/XYZSwapPair.sol

```diff
@@ -209,46 +280,59 @@ contract XYZSwapPair is IXYZSwapPair, ERC20Permit, ReentrancyGuard, VolumeTrendR
     }
 
     function verifyBalanceAndUpdateEma(
-        uint256 _amount0In,
-        uint256 _amount1In,
-        uint256 _reserve0,
-        uint256 _reserve1,
-        uint256 _balance0,
-        uint256 _balance1
+        uint256 amount0In,
+        uint256 amount1In,
+        uint256 beforeReserve0,
+        uint256 beforeReserve1,
+        uint256 afterReserve0,
+        uint256 afterReserve1
     ) internal virtual returns (uint256 feeInPrecision) {
-        // volume will be normalized into amount in token 0
-        uint256 volume = _reserve0.mul(_amount1In).div(_reserve1).add(_amount0In);
+        // volume = beforeReserve0 * amount1In / beforeReserve1 + amount0In (normalized into amount in token 0)
+        uint256 volume = beforeReserve0.mul(amount1In).div(beforeReserve1).add(amount0In);
         uint256 rFactorInPrecision = recordNewUpdatedVolume(block.number, volume);
         feeInPrecision = FeeFomula.getFee(rFactorInPrecision);
         //verify balance update is match with fomula
-        uint256 balance0Adjusted = _balance0.mul(PRECISION).sub(_amount0In.mul(feeInPrecision));
-        balance0Adjusted = balance0Adjusted.div(PRECISION);
-        uint256 balance1Adjusted = _balance1.mul(PRECISION).sub(_amount1In.mul(feeInPrecision));
-        balance1Adjusted = balance1Adjusted.div(PRECISION);
-        require(balance0Adjusted.mul(balance1Adjusted) >= _reserve0.mul(_reserve1), "XYZSwap: K");
+        uint256 balance0Adjusted = afterReserve0.mul(PRECISION);
+        balance0Adjusted = balance0Adjusted.sub(amount0In.mul(feeInPrecision));
+        balance0Adjusted = balance0Adjusted / PRECISION;
```

Comment:
`div` has been replaced with `/`. This is intentional right? It should be ok since it's safe multiplied by PRECISION before this line.

Reply:
Yeah. So for overflow, underflow, I follow standard that we can remove SafeMath if calculation is **explicit** safe from overflow-underflow.
The standard includes:
- divide by a variable which is not 0 (include constant != 0, totalSupply)
- subtract by a smaller uint. For example, if the condition `a>b` is checked above then `a.sub(b)` can replaced by `a-b`
- ranged safety. Any assumption about range should be checked inside a single pure function.

Note that: when we do Katatyst version, the range assumption is ambused. So it makes SafeMath review too complex.

Reply:
ok